acceptance-tests/drain_test.go

```diff
@@ -126,4 +135,68 @@ var _ = Describe("Drain Test", func() {
 		By("Sending a request to HAProxy works")
 		expectTestServer200(http.Get(fmt.Sprintf("http://%s", haproxyInfo.PublicIP)))
 	})
+
+	It("Closes idle connections gracefully", func() {
+		haproxyBackendPort := 12000
+		// Expect initial deployment to be failing due to lack of healthy backends
+		haproxyInfo, _ := deployHAProxy(baseManifestVars{
+			haproxyBackendPort:    haproxyBackendPort,
+			haproxyBackendServers: []string{"127.0.0.1"},
+			deploymentName:        deploymentNameForTestNode(),
+		}, []string{opsfileDrainTimeout}, map[string]interface{}{}, false)
+
+		// Verify that is in a failing state
+		Expect(boshInstances(deploymentNameForTestNode())[0].ProcessState).To(Or(Equal("failing"), Equal("unresponsive agent")))
+
+		closeLocalServer, localPort := startDefaultTestServer()
+		defer closeLocalServer()
+
+		closeTunnel := setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
+		defer closeTunnel()
+
+		By("Waiting monit to report HAProxy is now healthy (due to having a healthy backend instance)")
+		// Since the backend is now listening, HAProxy healthcheck should start returning healthy
+		// and monit should in turn start reporting a healthy process
+		// We will up to wait one minute for the status to stabilise
+		Eventually(func() string {
+			return boshInstances(deploymentNameForTestNode())[0].ProcessState
+		}, time.Minute, time.Second).Should(Equal("running"))
+
+		By("The healthcheck health endpoint should report a 200 status code")
+		expect200(http.Get(fmt.Sprintf("http://%s:8080/health", haproxyInfo.PublicIP)))
+
+		By("Sending request using keep-alive works")
+		conn, err := net.Dial("tcp", fmt.Sprintf("%s:80", haproxyInfo.PublicIP))
+		defer conn.Close()
+		Expect(err).ToNot(HaveOccurred())
+
+		sendHTTP := func(conn net.Conn) string {
+			_, err := conn.Write([]byte(strings.Join([]string{
+				"GET / HTTP/1.1",
+				fmt.Sprintf("Host: %s", haproxyInfo.PublicIP),
+				"Content-Length: 0",
+				"Content-Type: text/plain",
+				"\r\n",
+			}, "\r\n")))
+
+			Expect(err).ToNot(HaveOccurred())
+
+			// Too lazy to properly parse headers? Just stop reading after a second!
+			_ = conn.SetReadDeadline(time.Now().Add(time.Second))
+
+			response, _ := io.ReadAll(conn)
+			return string(response)
+		}
```

Comment:
The issue is that the Go HTTP client is smart enough to filter out things like "connection: close" and act upon them (e.g. close the connection). Plus it is relatively hard to get from the HTTP client down to the underlying TCP connection again and see if it is actually closed or not. 
So, to avoid these problems I chose to use a low-level approach to just compare the text output of the HTTP response and see if it contains the "connection: close" header.